### PR TITLE
Add polkadot support

### PR DESCRIPTION
**Pull request by jonasW3F**
I added Polkadot support based on the framework that was already present for Kusama. It definitely needs some more testing.

### CODE REVIEW

File: frontend/app/src/data/defaults.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ export class Defaults {
   static FLOATING_PRECISION = 2;
   static RPC_ENDPOINT = 'http://localhost:8545';
   static KSM_RPC_ENDPOINT = 'http://localhost:9933';
+  static DOT_RPC_ENDPOINT = 'http://localhost:9933'
```
Comment (yabirgb) [inline frontend/app/src/data/defaults.ts:11]: Here it's missing a `;`

Comment (yabirgb) [reply frontend/app/src/data/defaults.ts:11]: Also I don't know how this work but having the same port seems to be an issue

File: rotkehlchen/chain/manager.py [hunk 2]
```diff
@@ -52,7 +52,9 @@
 from rotkehlchen.chain.substrate.manager import wait_until_a_node_is_available
 from rotkehlchen.chain.substrate.typing import KusamaAddress
 from rotkehlchen.chain.substrate.utils import KUSAMA_NODE_CONNECTION_TIMEOUT
-from rotkehlchen.constants.assets import A_ADX, A_BTC, A_DAI, A_ETH, A_ETH2, A_KSM
+from rotkehlchen.chain.substrate.typing import PolkadotAddress
```
Comment (yabirgb) [inline rotkehlchen/chain/manager.py:55]: You have two imports from `from rotkehlchen.chain.substrate.typing` that can be combined

```suggestion
from rotkehlchen.chain.substrate.typing import KusamaAddress, PolkadotAddress
```

File: rotkehlchen/data/all_assets.json [hunk 3]
```diff
@@ -7078,6 +7078,13 @@
         "symbol": "KSM",
         "type": "own chain"
     },
+    "DOT": {
+        "coingecko": "polkadot",
+        "name": "Polkadot",
+        "started": 1590500178,
+        "symbol": "DOT",
+        "type": "own chain"
+    },
```
Comment (yabirgb) [inline rotkehlchen/data/all_assets.json:7087]: We don't use this file anymore. It's here for compatibility reasons. This change should be reverted and the asset added to one update in https://github.com/rotki/assets I'll add this to an issue

File: rotkehlchen/db/settings.py [hunk 4]
```diff
@@ -80,6 +80,7 @@ class DBSettings(NamedTuple):
     include_gas_costs: bool = DEFAULT_INCLUDE_GAS_COSTS
     eth_rpc_endpoint: str = 'http://localhost:8545'
     ksm_rpc_endpoint: str = 'http://localhost:9933'
+    dot_rpc_endpoint: str = 'http://localhost:9933'
```
Comment (yabirgb) [inline rotkehlchen/db/settings.py:83]: As I said I believe those having the same port could be a problem. What if you want both kusama and polkadot?

File: rotkehlchen/chain/manager.py (hunk 2 above)
Comment (jonasW3F) [reply rotkehlchen/chain/manager.py:55]: yep that was before I ran the lint. I added some commits

File: frontend/app/src/data/defaults.ts (hunk 1 above)
Comment (jonasW3F) [reply frontend/app/src/data/defaults.ts]: Let me find out if they have different ports

File: rotkehlchen/db/settings.py (hunk 4 above)
Comment (jonasW3F) [reply rotkehlchen/db/settings.py:83]: mh both seem to run on the same port by default. How could we fix this issue?

Comment (yabirgb) [reply rotkehlchen/db/settings.py:83]: What I would do here is store ports for kusama and polkadot as settings for the user in the database and make and option in the UI to change them. The default value could be this `9933`. But maybe @LefterisJP has a better idea?

Comment (yabirgb) [reply rotkehlchen/db/settings.py:83]: We have talked about this in our standup @jonasW3F . The default value for the port should be different and there is an option to set it already in the front-end, you should allow the customization of the port there

File: rotkehlchen/api/v1/encoding.py [hunk 5]
```diff
@@ -1453,6 +1476,59 @@ def _transform_ksm_address(
     return address
 
 
+def _transform_dot_address(
```
Comment (yabirgb) [inline rotkehlchen/api/v1/encoding.py:1479]: It might be a good idea to combine this function and `_transform_ksm_address` using a new argument `chain` and two `if`s statements inside the function to avoid duplication

File: rotkehlchen/chain/substrate/typing.py [hunk 6]
```diff
@@ -40,15 +41,52 @@ def endpoint(self) -> str:
                 'via "own_rpc_endpoint" or the specific db setting',
             )
         if self == KusamaNodeName.PARITY:
-            return 'https://kusama-rpc.polkadot.io/'
```
Comment (yabirgb) [inline rotkehlchen/chain/substrate/typing.py:43]: Why this change?

File: rotkehlchen/chain/substrate/typing.py [hunk 7]
```diff
@@ -59,12 +97,17 @@ class SubstrateChain(Enum):
...
     """
     KUSAMA = 1
+    POLKADOT = 2
 
     def __str__(self) -> SubstrateChainId:
         """Return the official chain identifier/name.
         """
         if self == SubstrateChain.KUSAMA:
             return SubstrateChainId('Kusama')
+
```
Comment (yabirgb) [inline rotkehlchen/chain/substrate/typing.py:107]: Unneeded space
```suggestion
```

File: rotkehlchen/chain/substrate/typing.py [hunk 8]
```diff
@@ -75,27 +118,42 @@ def chain_explorer_api(self) -> str:
         """
         if self == SubstrateChain.KUSAMA:
             return 'https://kusama.api.subscan.io/api'
+
```
Comment (yabirgb) [inline rotkehlchen/chain/substrate/typing.py:121]: Remove this space

File: rotkehlchen/chain/substrate/typing.py [hunk 9]
```diff
@@ -75,27 +118,42 @@ def chain_explorer_api(self) -> str:
...
+            return 'https://polkadot.api.subscan.io/api'
+
         raise AssertionError(f'Unexpected Chain: {self}')
 
     def substrate_interface_attributes(self) -> SubstrateInterfaceAttributes:
         """Return the attributes for instantiating SubstrateInterface.
         """
         if self == SubstrateChain.KUSAMA:
             return SubstrateInterfaceAttributes(type_registry_preset='kusama')
+
```
Comment (yabirgb) [inline rotkehlchen/chain/substrate/typing.py:132]: And this space. There are two more in this file

File: rotkehlchen/chain/substrate/typing.py [hunk 10]
```diff
@@ -75,27 +118,42 @@ def chain_explorer_api(self) -> str:
...
         raise AssertionError(f'Unexpected Chain: {self}')
 
     def blocks_threshold(self) -> BlockNumber:
         """Return the blocks difference that marks a node as unsynced.
         """
         if self == SubstrateChain.KUSAMA:
             return BlockNumber(10)
+
+        elif self == SubstrateChain.POLKADOT:
+            return BlockNumber(10)
```
Comment (yabirgb) [inline rotkehlchen/chain/substrate/typing.py:145]: Is the same number in both chains?

**Review by yabirgb — COMMENTED** [review]
Made a deeper review and had some comments + one question

File: rotkehlchen/chain/substrate/typing.py (hunk 6 above)
Comment (jonasW3F) [reply rotkehlchen/chain/substrate/typing.py:43]: changed it for testing and forgot to change back

File: rotkehlchen/chain/substrate/typing.py (hunk 10 above)
Comment (jonasW3F) [reply rotkehlchen/chain/substrate/typing.py:145]: Both Polkadot and Kusama have a 6 seconds block time, so I'd suggest sticking with the default that was set to Kusama.

Comment (yabirgb) [reply rotkehlchen/chain/substrate/typing.py:145]: Then I believe you can do something like 
```suggestion
 if self in (SubstrateChain.KUSAMA, SubstrateChain.POLKADOT):
            return BlockNumber(10)
```

File: rotkehlchen/api/v1/encoding.py (hunk 5 above)
Comment (jonasW3F) [reply rotkehlchen/api/v1/encoding.py]: I tried to integrate both functions but I am running into typing issues. Before, there was a `KusamaAddress` returned. Now, with combining the functions, I try to return a `SubstrateAddress` which already was defined as a Union of of `KusamaAddress` and `PolkadotAddress`

There is an error when trying to add a Polkadot Address now which points to the type error. Could you check? I spent quite some time but couldn't figure out a solution.

File: rotkehlchen/chain/substrate/typing.py (hunk 10 above)
Comment (jonasW3F) [reply rotkehlchen/chain/substrate/typing.py:142]: that's more elegant :) thanks

File: rotkehlchen/api/v1/encoding.py (hunk 5 above)
Comment (yabirgb) [reply rotkehlchen/api/v1/encoding.py]: @jonasW3F Okey Jonas. Will take a look today/tomorrow and see how this can be done :)

File: rotkehlchen/db/settings.py (hunk 4 above)
Comment (jonasW3F) [reply rotkehlchen/db/settings.py:83]: I think it is currently possible to set the ports in the front-end for both endpoints. How do we want to handle it? Do we only want to give one default for either KSM or DOT and leave the other empty?

File: rotkehlchen/api/v1/encoding.py (hunk 5 above)
Comment (yabirgb) [reply rotkehlchen/api/v1/encoding.py]: This would do the job. I've noticed that you use double quotes instead of single. In our style we use always single `'` to declare strings. Also notice the `Literal` on the type of the chain argument so the linter knows that this function should only be called using this strings.

If your run the linter `make lint` you will see several things to correct regarding to imports mostly.

Finally there is a file conflict in the PR. Try to rebase and solve the conflict

```
def _transform_substrate_address(
        ethereum: EthereumManager,
        given_address: str,
        chain: Literal['Kusama', 'Polkadot'],
) -> SubstrateAddress:
    """Returns a DOT or KSM address (if exists) given an ENS domain. At this point any
    given address has been already validated either as an ENS name or as a
    valid Polkadot address (ss58 format).

    NB: ENS domains for Substrate chains (e.g. KSM, DOT) store the Substrate
    public key. It requires to encode it with a specific ss58 format for
    obtaining the specific chain address.

    Polkadot/Polkadot ENS domain accounts:
    https://guide.Polkadot.network/docs/en/mirror-ens

    ENS domain substrate public key encoding:
    https://github.com/ensdomains/address-encoder/blob/master/src/index.ts
    """
    if not given_address.endswith('.eth'):
        if chain == 'Polkadot':
            return PolkadotAddress(given_address)
        elif chain == 'Kusama':
            return KusamaAddress(given_address)

    try:
        if(chain == 'Polkadot'):
            resolved_address = ethereum.ens_lookup(
                given_address,
                blockchain=SupportedBlockchain.POLKADOT,
            )
        if(chain == 'Kusama'):
            resolved_address = ethereum.ens_lookup(
                given_address,
                blockchain=SupportedBlockchain.KUSAMA,
            )
    except (RemoteError, InputError) as e:
        raise ValidationError(
            f'Given ENS address {given_address} could not be resolved '
            f'for ' + chain + ' due to: {str(e)}',
            field_name='address',
        ) from None

    if resolved_address is None:
        raise ValidationError(
            f'Given ENS address {given_address} could not be resolved for ' + chain,
            field_name='address',
        ) from None

    try:
        if(chain == 'Polkadot'):
            address = get_polkadot_address_from_public_key(SubstratePublicKey(resolved_address))
            log.debug(f'Resolved polkadot ENS {given_address} to {address}')
            return PolkadotAddress(address)
        if(chain == 'Kusama'):
            address = get_kusama_address_from_public_key(SubstratePublicKey(resolved_address))
            log.debug(f'Resolved kusama ENS {given_address} to {address}')
            return KusamaAddress(address)
    except (TypeError, ValueError) as e:
        raise ValidationError(
            f'Given ENS address {given_address} does not contain a valid '
            f'Substrate public key: {resolved_address}. ' + chain + ' address cannot be obtained.',
            field_name='address',
        ) from e
```

File: rotkehlchen/chain/substrate/typing.py (hunk 10 above)
Comment (jonasW3F) [reply rotkehlchen/chain/substrate/typing.py:142]: fixed in https://github.com/rotki/rotki/pull/3197/commits/18bed4103906e5e7c81298f91c8abf3a3f242088 thanks!

**Comment by LefterisJP** [comment]
Hey @jonasW3F when will this be ready for also a look by me? I see it also needs a rebase on develop.

Additionally you have a lot of commits in the PR. Can you squash together those that are related and keep only commits that do one thing together? For example, the fix, fix x, fix y commits can be squashed.

**Comment by jonasW3F** [comment]
> Hey @jonasW3F when will this be ready for also a look by me? I see it also needs a rebase on develop.
> 
> Additionally you have a lot of commits in the PR. Can you squash together those that are related and keep only commits that do one thing together? For example, the fix, fix x, fix y commits can be squashed.

Hi @LefterisJP 

I am currently swamped at work, sorry for being slow here. I'll work on it as soon as I can. I quickly tried the suggestion of @yabirgb on how to implement the `_transform_substrate_address` function but it still gave me the same error, that the given ENS address did not contain a valid substrate public key (I tried to add "bruno.eth" for example).

As soon as I find time, I also do a rebase and reduce the commits.

Cheers
Jonas

**Comment by LefterisJP** [comment]
> > Hey @jonasW3F when will this be ready for also a look by me? I see it also needs a rebase on develop.
> > Additionally you have a lot of commits in the PR. Can you squash together those that are related and keep only commits that do one thing together? For example, the fix, fix x, fix y commits can be squashed.
> 
> Hi @LefterisJP
> 
> I am currently swamped at work, sorry for being slow here. I'll work on it as soon as I can. I quickly tried the suggestion of @yabirgb on how to implement the `_transform_substrate_address` function but it still gave me the same error, that the given ENS address did not contain a valid substrate public key (I tried to add "bruno.eth" for example).
> 
> As soon as I find time, I also do a rebase and reduce the commits.
> 
> Cheers
> Jonas

Hey Jonas, understand. We can also try to help by doing some of this if you are too swamped. Just let us know.

File: rotkehlchen/data/all_assets.json (hunk 3 above)
Comment (jonasW3F) [reply rotkehlchen/data/all_assets.json]: Okay, I removed it.

File: rotkehlchen/db/settings.py (hunk 4 above)
Comment (jonasW3F) [reply rotkehlchen/db/settings.py:83]: I now leave the `dot_rpc_endpoint` blank so that a user can set it in the frontend manually. There might be better ways.

**Comment by LefterisJP** [comment]
Rebased on top of develop that contains the avalanche changes that caused so many conflicts. Also squashed a lot of relevant commits together.

Fixed frontend linting, backend flake8 linting but left mypy and pylint linting out for now since I would like to review it from the github interface first.

**Comment by LefterisJP** [comment]
@jonasW3F Since we have a lot of upcoming changes and we need to merge things to avoid a ton of conflicts again, I will try to take on the PR and push it. You also mentioned you are busy. I will ask you if I have any questions. I will use the PR review system to also note what should change, but will probably just do it on a commit on my own.

Please let me know if that's in some way not okay with you!

**Review by LefterisJP — COMMENTED** [review]
These are the issues I found. As discussed in discord will fix all of them and also fix some more problems after the rebase and when all green merge.

In another PR we would also need to add tests here.

**Review by LefterisJP — APPROVED** [review]
The 2 failing tests were timeouts and pass locally.

The lgmt comment is a false positive that is really annoying and happens often and does not even seem to be disabled with the proper comment.

Looks good and seems to work after the changes I added on top.

Thank you @jonasW3F. If you can find some time writing some tests in another PR would be really cool.

Otherwise anything else you may wanna add to rotki we really welcome it. Contributors like you is what keeps this project alive and kicking! :bow:

**Comment by jonasW3F** [comment]
Hi @LefterisJP 

Thank you very much for taking over and finishing the PR. Again, sorry for not delivering a polished PR :) I hope to find the time to write some tests soon, but again, feel free to take over.

Rotki is a great product, keep working on it :)

Cheers
Jonas
